lib/cli/src/wasmer_env.rs

```diff
@@ -0,0 +1,127 @@
+use std::path::{Path, PathBuf};
+
+use anyhow::{Context, Error};
+use once_cell::sync::Lazy;
+use url::Url;
+use wasmer_registry::WasmerConfig;
+
+/// Command-line flags for determining the local "Wasmer Environment".
+///
+/// This is where you access `$WASMER_DIR`, the `$WASMER_DIR/wasmer.toml` config
+/// file, and specify the current registry.
+#[derive(Debug, Clone, PartialEq, clap::Parser)]
+pub struct WasmerEnv {
+    /// Set Wasmer's home directory
+    #[clap(long, env = "WASMER_DIR", default_value = WASMER_DIR.as_os_str())]
+    wasmer_dir: PathBuf,
+    /// The registry to fetch packages from (inferred from the environment by
+    /// default)
+    #[clap(long, env = "WASMER_REGISTRY")]
+    registry: Option<Registry>,
+    /// The API token to use when communicating with the registry (inferred from
+    /// the environment by default)
+    #[clap(long, env = "WASMER_TOKEN")]
+    token: Option<String>,
+}
+
+impl WasmerEnv {
+    pub(crate) fn new(
```

Comment:
nit: what's the point of a constructor that just takes all three fields and puts them in the struct unmodified?

Reply:
The `Login` command needs to construct a `WasmerEnv` manually because its command-line flags aren't compatible with the `#[flatten] wasmer_env: WasmerEnv` approach.

I'd prefer to not make `WasmerEnv`'s fields public just so `Login` can construct a `WasmerEnv` directly. We want other sub-commands to go through`WasmerEnv`'s accessor methods so we always fall back to values from the `~/.wasmer/wasmer.toml` config.